posthog/helpers/session_recording.py

```diff
@@ -118,74 +117,187 @@ def decompress(base64data: str) -> str:
     return gzip.decompress(compressed_bytes).decode("utf-16", "surrogatepass")
 
 
-def paginate_snapshot_list(list_to_paginate: List, limit: Optional[int], offset: int) -> PaginatedSnapshotList:
-    if not limit:
-        has_next = False
-        paginated_list = list_to_paginate[offset:]
-    elif offset + limit < len(list_to_paginate):
-        has_next = True
-        paginated_list = list_to_paginate[offset : offset + limit]
-    else:
-        has_next = False
-        paginated_list = list_to_paginate[offset:]
-    return PaginatedSnapshotList(has_next=has_next, paginated_list=paginated_list)
-
-
-def paginate_chunk_decompression(
+def decompress_chunked_snapshot_data(
     team_id: int,
     session_recording_id: str,
-    all_recording_snapshots: List[SnapshotData],
+    all_recording_events: List[SnapshotDataTaggedWithWindowId],
     limit: Optional[int] = None,
     offset: int = 0,
-) -> PaginatedSnapshotList:
-    if len(all_recording_snapshots) == 0:
-        return PaginatedSnapshotList(has_next=False, paginated_list=[])
+    return_only_activity_data: bool = False,
+) -> DecompressedRecordingData:
+    """
+    Before data is stored in clickhouse, it is compressed and then chunked. This function
+    gets back to the original data by unchunking the events and then decompressing the data.
+    
+    If limit + offset is provided, then it will paginate the decompression by chunks (not by events, because 
+    you can't decompress an incomplete chunk).
+
+    Depending on the size of the recording, this function can return a lot of data. To decrease the
+    memory used, you should either use the pagination parameters or pass in 'return_only_activity_data' which
+    drastically reduces the size of the data returned if you only want the activity data (used for metadata calculation)
+    """
+
+    if len(all_recording_events) == 0:
+        return DecompressedRecordingData(has_next=False, snapshot_data_by_window_id={})
+
+    snapshot_data_by_window_id = defaultdict(list)
+
+    # Handle backward compatibility to the days of uncompressed and unchunked snapshots
+    if "chunk_id" not in all_recording_events[0].snapshot_data:
+        paginated_list = paginate_list(all_recording_events, limit, offset)
+        for event in paginated_list.paginated_list:
+            snapshot_data_by_window_id[event.window_id].append(event.snapshot_data)
```

Comment:
It could be `None`, but it's taking advantage of python allowing for `None` to be a key in a dict. So basically, all events that don't have a `window_id` get grouped together.